File: ruby/testcode/fileupload_004.rb
require_relative 'shared'

SAFE_EXTENSIONS = %w[jpg png gif pdf].freeze

# vuln-code-snippet start ruby_fileupload_uuid_rename
def upload_renamed(req)
  upload = req.file('file')
  ext = File.extname(upload[:filename]).delete('.').downcase
  return BenchmarkResponse.bad_request('not allowed') unless SAFE_EXTENSIONS.include?(ext)
  safe_name = SecureRandom.uuid + '.' + ext # vuln-code-snippet safe-line ruby_fileupload_uuid_rename
  File.write("/uploads/#{safe_name}", upload[:data])
  BenchmarkResponse.ok(safe_name)
end
# vuln-code-snippet end ruby_fileupload_uuid_rename

File: rust/testcode/fileupload_004.rs
//! CWE-434: Upload destination inside web-servable static directory.

// vuln-code-snippet start testcodeFileupload004
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let filename = req.param("filename");
    let content = req.body_str();
    let path = format!("./static/uploads/{}", filename); // vuln-code-snippet target-line testcodeFileupload004
    let _ = std::fs::write(&path, content.as_bytes());
    super::shared::BenchmarkResponse::ok(&format!("Saved: {}", path))
}
// vuln-code-snippet end testcodeFileupload004
